### PR TITLE
Extend builder support for quantized lm_head

### PR DESCRIPTION
**Pull request by BowenBao**
As title. This enables support for further reduced quantized model size and improved runtime efficiency, within acceptable range of accuracy degradation.

Orthogonal to #940. This PR targets already quantized models in autoawq/autogptq format that **has** lmhead quantized.

### CODE REVIEW

**Comment by BowenBao** [comment]
@kunal-vaishnavi ptal, thanks.

File: src/python/py/models/quantized_model.py [hunk 1]
```diff
@@ -339,10 +351,39 @@ def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, in
         # Set properties of each layer based on quantization type
         self.set_properties()
 
+    def _initialize_quantized_lm_head(self, bits, group_size):
+        """
+        Initialize `QuantizedTensorModule` for LM head if not already set
+        """
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py:357]: Should we consider replacing `TensorModule` with `QuantizedTensorModule`, moving `self.weight` and `self.bias` to `QuantizedTensorModule`, and just use `QuantizedTensorModule`? Then, we wouldn't need to initialize and we can still check between quantized LM head vs. unquantized LM head with `self.lm.qweight is not None`.

File: src/python/py/models/quantized_model.py [hunk 2]
```diff
@@ -581,6 +622,12 @@ def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, in
 
                     # Set `g_idx` to None since it's not used in `MatMulNBits`
                     q_tensors.g_idx = None
+        if isinstance(self.lm_head, QuantizedTensorModule) and self.lm_head.qweight is not None:
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py:625]:
```suggestion

        if isinstance(self.lm_head, QuantizedTensorModule) and self.lm_head.qweight is not None:
```

File: src/python/py/models/quantized_model.py [hunk 3]
```diff
@@ -339,10 +351,39 @@ def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, in
...
+                raise NotImplementedError(
+                    "lm_head quantization is only supported for awq."
+                )
+            self.lm_head.out_features = self.lm_head.scales.shape[1]
+            self.lm_head.in_features = self.lm_head.qweight.shape[0]
+            # Set g_idx if not already set
+            self.lm_head.g_idx = (
+                self.lm_head.g_idx
+                if self.lm_head.g_idx is not None
+                else torch.tensor(
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py:379]: Can this be in one line?

File: src/python/py/models/quantized_model.py [hunk 4]
```diff
@@ -339,10 +351,39 @@ def __init__(self, quant_type, input_path, bits, group_size, q_size, kv_size, in
...
+            self.lm_head = QuantizedTensorModule(bits, group_size)
+
     def set_properties(self):
         """
         Set in_features, out_features, and g_idx based on quantization type
         """
+        if isinstance(self.lm_head, QuantizedTensorModule):
+            if self.quant_type != "awq":
+                raise NotImplementedError(
+                    "lm_head quantization is only supported for awq."
```
Comment (kunal-vaishnavi) [inline src/python/py/models/quantized_model.py:371]: Can we set the properties for GPTQ-quantized LM head as well? It should be similar to setting the properties for AWQ-quantized LM head.

**Review by kunal-vaishnavi — APPROVED** [review]
